pennylane/qaoa/layers.py

```diff
@@ -0,0 +1,165 @@
+        ValueError: if the terms of the supplied cost Hamiltonian are not exclusively products of diagonal Pauli gates
+
+    .. UsageDetails::
+
+        We define a cost Hamiltonian and pass it into ``cost_layer``:
+        and pass it into ``cost_layer``:
+
+        .. code-block:: python
+
+            from pennylane import qaoa
+            import pennylane as qml
+
+            cost_h = qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliZ(1)])
+            cost_layer = qaoa.cost_layer(cost_h)
+
+        We can then use the cost layer within a quantum circuit:
+
+        .. code-block:: python
+
+            dev = qml.device('default.qubit', wires=2)
+
+            @qml.qnode(dev)
+            def circuit(gamma):
+
+                for i in range(2):
+                    qml.Hadamard(wires=i)
+
+                cost_layer(gamma)
+
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(2)]
+
+        which gives us a circuit of the form:
+
+        >>> circuit(0.5)
+
+        .. code-block:: none
+
+             0: ──H──RZ(-1.0)──╭RZ(-1.0)──┤ ⟨Z⟩
+             1: ──H────────────╰RZ(-1.0)──┤ ⟨Z⟩
+
+    """
+    if not isinstance(hamiltonian, qml.Hamiltonian):
+        raise ValueError(
+            "hamiltonian must be of type pennylane.Hamiltonian, got {}".format(
+                type(hamiltonian).__name__
+            )
+        )
+
+    if not _diagonal_terms(hamiltonian):
+        raise ValueError("hamiltonian must be written only in terms of PauliZ and Identity gates")
+
+    return lambda gamma: qml.templates.ApproxTimeEvolution(hamiltonian, gamma, 1)
+
+
+def mixer_layer(hamiltonian):
+    r"""Returns the QAOA cost layer corresponding to a mixer Hamiltonian.
+
+    For a mixer Hamiltonian :math:`H_M`, this is defined as the following unitary:
+
+    .. math:: U_M \ = \ e^{-i \alpha H_M}
+
+    where :math:`\alpha` is a variational parameter.
+
+    Args:
+        hamiltonian (.Hamiltonian): The mixer Hamiltonian
+
+    .. UsageDetails::
+
+        We define a mixer Hamiltonian and pass it into ``mixer_layer``:
+
+        .. code-block:: python
+
+            from pennylane import qaoa
+            import pennylane as qml
+
+            mixer_h = qml.Hamiltonian([1, 1], [qml.PauliX(0), qml.PauliX(0) @ qml.PauliX(1)])
+            mixer_layer = qaoa.mixer_layer(mixer_h)
+
+        We can then use the cost layer within a quantum circuit:
+
+        .. code-block:: python
+
+            dev = qml.device('default.qubit', wires=2)
+
+            @qml.qnode(dev)
+            def circuit(alpha):
+
+                for i in range(2):
+                    qml.Hadamard(wires=i)
+
+                mixer_layer(alpha)
+
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(2)]
+
+        which gives us a circuit of the form:
+
+        >>> circuit(0.5)
+
+        .. code-block:: none
+
+             0: ──H──RZ(-1.0)──H──H──╭RZ(-1.0)──H──┤ ⟨Z⟩
+             1: ──H──────────────────╰RZ(-1.0)──H──┤ ⟨Z⟩
```

Comment:
Suggest making the same changes here regarding having a single console code block